### PR TITLE
CSHARP-4698: Expression not supported: Regex.IsMatch

### PR DESCRIPTION
**Pull request by sanych-sun**

### CODE REVIEW

File: tests/MongoDB.Driver.Tests/Linq/Linq3Implementation/Translators/ExpressionToAggregationExpressionTranslators/MethodTranslators/IsMatchMethodToAggregationExpressionTranslatorTests.cs [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+namespace MongoDB.Driver.Tests.Linq.Linq3Implementation.Translators.ExpressionToAggregationExpressionTranslators.MethodTranslators
+{
+    public class IsMatchMethodToAggregationExpressionTranslatorTests : Linq3IntegrationTest
+    {
+        [Fact]
+        public void Should_translate_instance_regex_isMatch()
+        {
+            var collection = CreateCollection();
+            var regex = new Regex(@"\dB.*0");
+            var query = collection.AsQueryable()
```
Comment (rstam) [inline tests/MongoDB.Driver.Tests/Linq/Linq3Implementation/Translators/ExpressionToAggregationExpressionTranslators/MethodTranslators/IsMatchMethodToAggregationExpressionTranslatorTests.cs:15]: Our convention in our tests is to name this variable `queryable`.

File: tests/MongoDB.Driver.Tests/Linq/Linq3Implementation/Translators/ExpressionToAggregationExpressionTranslators/MethodTranslators/IsMatchMethodToAggregationExpressionTranslatorTests.cs [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+                .Where(i => regex.IsMatch(i.A + i.B));
+
+            var translated = Translate(collection, query);
```
Comment (rstam) [inline tests/MongoDB.Driver.Tests/Linq/Linq3Implementation/Translators/ExpressionToAggregationExpressionTranslators/MethodTranslators/IsMatchMethodToAggregationExpressionTranslatorTests.cs:18]: Our convention in our tests is to name this variable `stages`.

File: tests/MongoDB.Driver.Tests/Linq/Linq3Implementation/Translators/ExpressionToAggregationExpressionTranslators/MethodTranslators/IsMatchMethodToAggregationExpressionTranslatorTests.cs [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+            AssertStages(translated, @"{ '$match' : { '$expr' : { '$regexMatch' : { 'input' : { '$concat' : ['$A', '$B'] }, 'regex' : '\\dB.*0', 'options' : '' } } } }");
+
+            var results = query.ToList();
+            results.Should().OnlyContain(i => Regex.IsMatch(i.A + i.B, @"\dB.*0"));
```
Comment (rstam) [inline tests/MongoDB.Driver.Tests/Linq/Linq3Implementation/Translators/ExpressionToAggregationExpressionTranslators/MethodTranslators/IsMatchMethodToAggregationExpressionTranslatorTests.cs:22]: I get concerned when I see "logic" in assertions. Consider this simpler assertion instead:

     results.Select(x => x.Id).Should().Equal(2);

File: tests/MongoDB.Driver.Tests/Linq/Linq3Implementation/Translators/ExpressionToAggregationExpressionTranslators/MethodTranslators/IsMatchMethodToAggregationExpressionTranslatorTests.cs [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+                new Data { Id = 1, A = "ABC", B = "1" },
+                new Data { Id = 2, A = "1Br", B = "0" });
+            return collection;
+        }
+
+        private class Data
+        {
+            public int Id { get; set; }
+            public string A { get; set; }
+
```
Comment (rstam) [inline tests/MongoDB.Driver.Tests/Linq/Linq3Implementation/Translators/ExpressionToAggregationExpressionTranslators/MethodTranslators/IsMatchMethodToAggregationExpressionTranslatorTests.cs:67]: Extra blank line

**Review by rstam — COMMENTED** [review]
Overall looks good. See minor changes requested or suggested.

File: src/MongoDB.Driver/Linq/Linq3Implementation/Reflection/RegexMethod.cs [hunk 5]
```diff
@@ -37,5 +39,61 @@ static RegexMethod()
...
         public static MethodInfo StaticIsMatch => __staticIsMatch;
         public static MethodInfo StaticIsMatchWithOptions => __staticIsMatchWithOptions;
+
+        // public methods
+        public static bool IsMatchMethod(MethodCallExpression expression, out Expression inputExpression, out Regex regex)
+        {
+            var method = expression.Method;
+            var arguments = expression.Arguments;
+
+            if (method.Is(IsMatch))
```
[the remark this answers is not in the document]
Comment (sanych-sun) [reply src/MongoDB.Driver/Linq/Linq3Implementation/Reflection/RegexMethod.cs:49]: Done

File: src/MongoDB.Driver/Linq/Linq3Implementation/Reflection/RegexMethod.cs [hunk 6]
```diff
@@ -37,5 +39,61 @@ static RegexMethod()
...
+                var objectExpression = expression.Object;
+                if (objectExpression is ConstantExpression objectConstantExpression)
+                {
+                    regex = (Regex)objectConstantExpression.Value;
+                    inputExpression = arguments[0];
+                    return true;
+                }
+            }
+
+            if (method.IsOneOf(StaticIsMatch, StaticIsMatchWithOptions))
```
[the remark this answers is not in the document]
Comment (sanych-sun) [reply src/MongoDB.Driver/Linq/Linq3Implementation/Reflection/RegexMethod.cs:60]: Done

File: src/MongoDB.Driver/Linq/Linq3Implementation/Reflection/RegexMethod.cs [hunk 7]
```diff
@@ -37,5 +39,61 @@ static RegexMethod()
...
+                }
+
+                regex = new Regex(pattern, options);
+                return true;
+            }
+
+            returnFalse:
+            inputExpression = null;
+            regex = null;
+            return false;
```
[the remark this answers is not in the document]
Comment (sanych-sun) [reply src/MongoDB.Driver/Linq/Linq3Implementation/Reflection/RegexMethod.cs:96]: Done

File: src/MongoDB.Driver/Linq/Linq3Implementation/Translators/ExpressionToAggregationExpressionTranslators/MethodTranslators/IsMatchMethodToAggregationExpressionTranslator.cs [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+using MongoDB.Driver.Linq.Linq3Implementation.Reflection;
+
+namespace MongoDB.Driver.Linq.Linq3Implementation.Translators.ExpressionToAggregationExpressionTranslators.MethodTranslators
+{
+    internal static class IsMatchMethodToAggregationExpressionTranslator
+    {
+        // public static methods
+        public static AggregationExpression Translate(TranslationContext context, MethodCallExpression expression)
+        {
+            if (RegexMethod.IsMatchMethod(expression, out var inputExpression, out var regularExpression))
```
[the remark this answers is not in the document]
Comment (sanych-sun) [reply src/MongoDB.Driver/Linq/Linq3Implementation/Translators/ExpressionToAggregationExpressionTranslators/MethodTranslators/IsMatchMethodToAggregationExpressionTranslator.cs:31]: Done

File: src/MongoDB.Driver/Linq/Linq3Implementation/Translators/ExpressionToAggregationExpressionTranslators/MethodTranslators/IsMatchMethodToAggregationExpressionTranslator.cs [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+
+        private static AggregationExpression Translate(TranslationContext context, Expression expression, Expression inputExpression, Regex regex)
+        {
+            var inputTranslation = ExpressionToAggregationExpressionTranslator.Translate(context, inputExpression);
+            var regularExpression = new BsonRegularExpression(regex);
+
+            if (inputTranslation.Serializer is IRepresentationConfigurable representationConfigurable &&
+                representationConfigurable.Representation != BsonType.String)
+            {
+                throw new ExpressionNotSupportedException(inputExpression, expression, because: $"input expression is not evaluated as a string");
```
[the remark this answers is not in the document]
Comment (sanych-sun) [reply src/MongoDB.Driver/Linq/Linq3Implementation/Translators/ExpressionToAggregationExpressionTranslators/MethodTranslators/IsMatchMethodToAggregationExpressionTranslator.cs:47]: Done

File: src/MongoDB.Driver/Linq/Linq3Implementation/Translators/ExpressionToAggregationExpressionTranslators/MethodTranslators/IsMatchMethodToAggregationExpressionTranslator.cs [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+            }
+
+            var ast = AstExpression.RegexMatch(inputTranslation.Ast, regularExpression.Pattern, regularExpression.Options);
+            return new AggregationExpression(expression, ast, new BooleanSerializer());
```
[the remark this answers is not in the document]
Comment (sanych-sun) [reply src/MongoDB.Driver/Linq/Linq3Implementation/Translators/ExpressionToAggregationExpressionTranslators/MethodTranslators/IsMatchMethodToAggregationExpressionTranslator.cs:51]: Done

File: src/MongoDB.Driver/Linq/Linq3Implementation/Translators/ExpressionToAggregationExpressionTranslators/MethodTranslators/IsMatchMethodToAggregationExpressionTranslator.cs [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+namespace MongoDB.Driver.Linq.Linq3Implementation.Translators.ExpressionToAggregationExpressionTranslators.MethodTranslators
+{
+    internal static class IsMatchMethodToAggregationExpressionTranslator
+    {
+        // public static methods
+        public static AggregationExpression Translate(TranslationContext context, MethodCallExpression expression)
+        {
+            if (RegexMethod.IsMatchMethod(expression, out var inputExpression, out var regularExpression))
+            {
+                return Translate(context, expression, inputExpression, regularExpression);
```
[the remark this answers is not in the document]
Comment (sanych-sun) [reply src/MongoDB.Driver/Linq/Linq3Implementation/Translators/ExpressionToAggregationExpressionTranslators/MethodTranslators/IsMatchMethodToAggregationExpressionTranslator.cs:33]: Done

File: src/MongoDB.Driver/Linq/Linq3Implementation/Translators/ExpressionToAggregationExpressionTranslators/MethodTranslators/IsMatchMethodToAggregationExpressionTranslator.cs [hunk 12]
```diff
@@ -0,0 +1,54 @@
...
+* limitations under the License.
+*/
+
+using System.Linq.Expressions;
+using System.Text.RegularExpressions;
+using MongoDB.Bson;
+using MongoDB.Bson.Serialization;
+using MongoDB.Bson.Serialization.Serializers;
+using MongoDB.Driver.Linq.Linq3Implementation.Ast.Expressions;
+using MongoDB.Driver.Linq.Linq3Implementation.Reflection;
```
[the remark this answers is not in the document]
Comment (sanych-sun) [reply src/MongoDB.Driver/Linq/Linq3Implementation/Translators/ExpressionToAggregationExpressionTranslators/MethodTranslators/IsMatchMethodToAggregationExpressionTranslator.cs:22]: Done

File: src/MongoDB.Driver/Linq/Linq3Implementation/Translators/ExpressionToFilterTranslators/MethodTranslators/IsMatchMethodToFilterTranslator.cs [hunk 13]
```diff
@@ -29,70 +28,14 @@ internal static class IsMatchMethodToFilterTranslator
         // public static methods
         public static AstFilter Translate(TranslationContext context, MethodCallExpression expression)
         {
-            if (IsMatchMethod(expression, out var inputExpression, out var regularExpression))
+            if (RegexMethod.IsMatchMethod(expression, out var inputExpression, out var regularExpression))
```
[the remark this answers is not in the document]
Comment (sanych-sun) [reply src/MongoDB.Driver/Linq/Linq3Implementation/Translators/ExpressionToFilterTranslators/MethodTranslators/IsMatchMethodToFilterTranslator.cs:31]: Done

File: tests/MongoDB.Driver.Tests/Linq/Linq3Implementation/Translators/ExpressionToAggregationExpressionTranslators/MethodTranslators/IsMatchMethodToAggregationExpressionTranslatorTests.cs [hunk 14]
```diff
@@ -0,0 +1,71 @@
+﻿using System.Text.RegularExpressions;
```
[the remark this answers is not in the document]
Comment (sanych-sun) [reply tests/MongoDB.Driver.Tests/Linq/Linq3Implementation/Translators/ExpressionToAggregationExpressionTranslators/MethodTranslators/IsMatchMethodToAggregationExpressionTranslatorTests.cs:1]: ;( Done

File: tests/MongoDB.Driver.Tests/Linq/Linq3Implementation/Translators/ExpressionToAggregationExpressionTranslators/MethodTranslators/IsMatchMethodToAggregationExpressionTranslatorTests.cs (hunk 1 above)
Comment (sanych-sun) [reply tests/MongoDB.Driver.Tests/Linq/Linq3Implementation/Translators/ExpressionToAggregationExpressionTranslators/MethodTranslators/IsMatchMethodToAggregationExpressionTranslatorTests.cs:19]: Done

File: tests/MongoDB.Driver.Tests/Linq/Linq3Implementation/Translators/ExpressionToAggregationExpressionTranslators/MethodTranslators/IsMatchMethodToAggregationExpressionTranslatorTests.cs (hunk 2 above)
Comment (sanych-sun) [reply tests/MongoDB.Driver.Tests/Linq/Linq3Implementation/Translators/ExpressionToAggregationExpressionTranslators/MethodTranslators/IsMatchMethodToAggregationExpressionTranslatorTests.cs:22]: Done

File: tests/MongoDB.Driver.Tests/Linq/Linq3Implementation/Translators/ExpressionToAggregationExpressionTranslators/MethodTranslators/IsMatchMethodToAggregationExpressionTranslatorTests.cs (hunk 4 above)
Comment (sanych-sun) [reply tests/MongoDB.Driver.Tests/Linq/Linq3Implementation/Translators/ExpressionToAggregationExpressionTranslators/MethodTranslators/IsMatchMethodToAggregationExpressionTranslatorTests.cs:75]: Done

File: tests/MongoDB.Driver.Tests/Linq/Linq3Implementation/Translators/ExpressionToAggregationExpressionTranslators/MethodTranslators/IsMatchMethodToAggregationExpressionTranslatorTests.cs (hunk 3 above)
Comment (sanych-sun) [reply tests/MongoDB.Driver.Tests/Linq/Linq3Implementation/Translators/ExpressionToAggregationExpressionTranslators/MethodTranslators/IsMatchMethodToAggregationExpressionTranslatorTests.cs:26]: Done

File: src/MongoDB.Driver.Core/Core/Misc/Feature.cs [hunk 15]
```diff
@@ -600,6 +601,12 @@ public class Feature
         [Obsolete("This property will be removed in a later release.")]
         public static Feature ReadConcern => __readConcern;
 
+
+        /// <summary>
+        /// Performs a regular expression pattern matching
```
Comment (rstam) [inline src/MongoDB.Driver.Core/Core/Misc/Feature.cs:606]: /// Gets the regex match feature.

File: src/MongoDB.Driver/Linq/Linq3Implementation/Translators/ExpressionToFilterTranslators/MethodTranslators/IsMatchMethodToFilterTranslator.cs [hunk 16]
```diff
@@ -29,70 +28,14 @@ internal static class IsMatchMethodToFilterTranslator
         // public static methods
         public static AstFilter Translate(TranslationContext context, MethodCallExpression expression)
         {
-            if (IsMatchMethod(expression, out var inputExpression, out var regularExpression))
+            if (RegexMethod.IsMatchMethod(expression, out var inputExpression, out var regex))
             {
-                return Translate(context, expression, inputExpression, regularExpression);
+                return Translate(context, expression, inputExpression, regex);
```
Comment (rstam) [inline src/MongoDB.Driver/Linq/Linq3Implementation/Translators/ExpressionToFilterTranslators/MethodTranslators/IsMatchMethodToFilterTranslator.cs:33]: Do you want to inline the `Translate` method here the same as we did in the `IsMatchMethodToAggregationExpressionTranslator` file?

**Review by rstam — CHANGES_REQUESTED** [review]

File: src/MongoDB.Driver.Core/Core/Misc/Feature.cs (hunk 15 above)
Comment (sanych-sun) [reply src/MongoDB.Driver.Core/Core/Misc/Feature.cs:606]: Done

File: src/MongoDB.Driver/Linq/Linq3Implementation/Translators/ExpressionToFilterTranslators/MethodTranslators/IsMatchMethodToFilterTranslator.cs (hunk 16 above)
Comment (sanych-sun) [reply src/MongoDB.Driver/Linq/Linq3Implementation/Translators/ExpressionToFilterTranslators/MethodTranslators/IsMatchMethodToFilterTranslator.cs]: Done

**Review by rstam — APPROVED** [review]
LGTM

**Pull request merged**